### ISSUE TITLE
Callback URL may not be parsed correctly by the regex

### ISSUE THREAD

**Issue by jeswin** [simov/grant#162]
The regex that is used to match /connect/:provider/callback doesn't seem to be handling some (but not all) Google accounts.

Here are a couple of different redirects issued by Google:

1. This works

```
http://example.com/connect/google/callback?state=e16a8...&code=4%2FzAHfquZ....
```

2. This does not work.

```
http://example.com/connect/google/callback?state=e16a8...&code=4/zAHfquZ....
```

The difference is that in the first case, the slash is encoded as %2F and it matches the regex defined.

From lib/handler/koa-2.js
```js
var regex = new RegExp([
    '^',
    app.config.defaults.prefix.replace(/\//g, '\/'),
    /\/([^\\/]+?)/.source, // /:provider
    /(?:\/([^\\/]+?))?/.source, // /:override?
    /\/?(?:\?([^/]+))?$/.source, // querystring
  ].join(''))
```

According to the implementation of the URL class in JavaScript it seems that everything after the question mark should be part of the querystring. Perhaps the regex above needs a fix?

**Comment by simov** [comment]
That's interesting because I used the internal module responsible for path matching in Express to generate this RegExp, so theoretically it should be correct.

In which case do you receive `code=4%2FzAHfquZ` and in which `code=4/zAHfquZ`?

**Comment by jeswin** [comment]
Unfortunately, I can't pin that down yet. When I use my google apps account, I see this problem. With a couple of other accounts I don't. I can investigate this a bit more. Also, Firefox can handle this without issues - while Chrome can't.

I can investigate futher. Thank you.

**Comment by simov** [comment]
You are correct there was a bug in the path matching regexp. Somehow I managed to add the wrong path matching regexp from a feature that I'm still developing. Obviously that means my tests doesn't cover that either which I'll be fixing pretty soon.

The fix was published in v5.0.1 let me know if it works for you.

**Comment by jeswin** [comment]
And it works! 

I want to thank you for this wonderful project, which is of significant help to thousands of developers. And for how politely and professionally you respond to users on these forums. More success to you, and stay safe.
